File: android/src/main/java/com/bluetrum/fota/OtaManager.java
package com.bluetrum.fota;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.bluetrum.devicemanager.DeviceCommManager;
import com.bluetrum.devicemanager.cmd.Request;
import com.bluetrum.fota.cmd.request.OtaInfoRequest;
import com.bluetrum.utils.CryptoUtils;

import java.nio.ByteBuffer;

public class OtaManager implements DeviceCommManager.NotificationCallback<Byte> {

    private static final String TAG = OtaManager.class.getSimpleName();

    private static final int UNDEFINED_FIRMWARE_VERSION = 0xFFFFFFFF;

    private static final int DEFAULT_TIMEOUT = 10000; // in ms, 10s
    private final Handler timeoutHandler;
    private final Handler notifyHandler;

    protected OtaDataProvider dataProvider;
    protected OtaRequestGenerator requestGenerator;

    protected int firmwareVersion = UNDEFINED_FIRMWARE_VERSION;
    protected byte[] hashData = new byte[0];
    protected int otaDataSize;

    protected boolean allowedUpdate;
    protected boolean isUpdating;
    protected boolean isUpdatePause;

    private final RequestDelegate requestDelegate;
    private final EventListener eventListener;

    private static final int MAX_RETRY_COUNT = 10;
    private int maxRetryCount = MAX_RETRY_COUNT;
    private int retryCount = 0;

    private RetryCallback retryCallback;

    public OtaManager(@NonNull RequestDelegate requestDelegate,
                      @NonNull EventListener eventListener) {
        this.requestDelegate = requestDelegate;
        this.eventListener = eventListener;

        this.timeoutHandler = new Handler(Looper.getMainLooper());
        this.notifyHandler = new Handler(Looper.getMainLooper());
        this.requestGenerator = new OtaRequestGenerator();
    }

    /**
     *  准备升级信息，需要先使用{@link OtaManager#setOtaData}传入OTA数据
     */
    public void prepareToUpdate() {
        if (dataProvider != null) {
            requireOtaInfo();
        }
    }

    /**
     * 开始OTA升级，内部需要allowUpdate，外部需要确保设备准备就绪（TWS连接等）
     */
    public void startOTA() {
        if (isReadyToUpdate()) {
            isUpdating = true;
            notifyOnStart();
            sendBlock();
        }
    }

    /**
     * {@link OtaManager}是否已经准备好升级
     * @return 是否已经准备好升级
     */
    public boolean isReadyToUpdate() {
        return dataProvider != null && allowedUpdate;
    }

    public void setRetryCallback(RetryCallback retryCallback) {
        this.retryCallback = retryCallback;
    }

    public void setMaxRetryCount(int maxRetryCount) {
        this.maxRetryCount = maxRetryCount;
    }

    private boolean isRetrying() {
        return retryCount != 0;
    }

    private boolean needRetryOnState(int state) {
        // Retry while in these states
        if (state == OtaError.ERROR_CODE_WRONG_SEQ_NUM
                || state == OtaError.ERROR_CODE_WRONG_DATA_LEN) {
            retryCount++;
            return retryCount <= maxRetryCount;
        }
        return false;
    }

    private void retryUpdate() {
        if (retryCallback != null) {
            retryCallback.beforeRetryCallback();
        }
        // Prepare to update, again
        prepareToUpdate();
        if (retryCallback != null) {
            retryCallback.afterRetryCallback();
        }
    }

    /**
     * 重置{@link OtaManager}，设备断开等情况调用
     */
    public void reset() {
//        firmwareVersion = UNDEFINED_FIRMWARE_VERSION;
//        hashData = new byte[0];
//        otaDataSize = 0;
        allowedUpdate = false;
        isUpdating = false;
        isUpdatePause = false;
        cancelTimeout();
    }

    /* 发送状态 */

    protected void notifyAllowUpdate(boolean allowed) {
        notifyHandler.post(() -> eventListener.onOtaAllowUpdate(allowed));
    }

    protected void notifyOnStart() {
        notifyHandler.post(eventListener::onOtaStart);
    }

    protected void notifyOnProgress(int progress) {
        notifyHandler.post(() -> eventListener.onOtaProgress(progress));
    }

    protected void notifyOnFinish() {
        notifyHandler.post(eventListener::onOtaFinish);
    }

    protected void notifyOnPause() {
        notifyHandler.post(eventListener::onOtaPause);
    }

    protected void notifyOnContinue() {
        notifyHandler.post(eventListener::onOtaContinue);
    }

    protected void notifyOnError(int errorCode) {
        notifyHandler.post(() -> eventListener.onOtaError(errorCode));
    }

    /* 发送指令 */

    protected void requireOtaInfo() {
        Request request = new OtaInfoRequest(firmwareVersion, hashData, otaDataSize);
        sendRequest(request, new DeviceCommManager.RequestCallback() {
            @Override
            public void onComplete(@NonNull Request request, @Nullable Object result) {
                if (result != null) {
                    OtaInfo otaInfo = (OtaInfo) result;
                    int startAddress = otaInfo.getStartAddress();
                    int blockSize = otaInfo.getBlockSize();
                    boolean allowUpdate = otaInfo.isAllowUpdate();

//                    Timber.v("startAddr = %s, blockSize = %s, allowUpdate = %s", startAddress, blockSize, allowUpdate);
                    Log.v(TAG, "startAddr = " + startAddress +", blockSize = " + blockSize +", allowUpdate = " + allowUpdate);

                    dataProvider.setStartAddress(startAddress);
                    dataProvider.setBlockSize(blockSize);
//                    otaManager.setPacketSize(commManager.getMaxPayloadSize()); // todo: app layer

                    allowedUpdate = allowUpdate;

                    // If it's retrying, start updating automatically
                    if (isRetrying()) {
                        startOTA();
                    } else if (isUpdatePause) { // 如果是从暂停中恢复，自动开始升级流程
                        isUpdatePause = false;
                        startOTA();
                    } else {
                        notifyAllowUpdate(allowUpdate);
                    }
                }
            }

            @Override
            public void onTimeout(@NonNull Request request) {
                // todo: notify timeout
            }
        });
    }

    protected void sendOtaStart() {
        Request request = requestGenerator.getOtaStartRequest();
        sendRequest(request);
        // 进度
        notifyOnProgress(dataProvider.getProgress());
    }

    protected void sendOtaData() {
        Request request = requestGenerator.getOtaSendDataRequest();
        sendRequest(request);
        // 进度
        notifyOnProgress(dataProvider.getProgress());
    }

    protected boolean canSendNow() {
        return !dataProvider.isBlockSentFinish();
    }

    protected void sendOtaDataOnce() {
        sendOtaData();
    }

    protected void doSendData() {
        while (canSendNow()) {
            sendOtaDataOnce();
        }
        // 发完块的最后一包后，就等待固件回复
        if (dataProvider != null && dataProvider.isBlockSentFinish()) {
            // 等待固件回复
            timeoutHandler.postDelayed(this::handleTimeout, DEFAULT_TIMEOUT);
        }
    }

    protected void sendBlock() {
        sendOtaStart();
        doSendData();
    }

    private void handleTimeout() {
        // 错误通知
//        notifyOnError(OtaError.TIMEOUT_RECEIVE_RESPONSE);
    }

    private void cancelTimeout() {
        timeoutHandler.removeCallbacksAndMessages(null);
    }

    private byte[] getHash(@NonNull byte[] otaData) {
        ByteBuffer bb = ByteBuffer.allocate(4);
        byte[] md5 = CryptoUtils.getMD5(otaData);
        if (md5 != null) {
            bb.put(md5, 0, 4);
        } else {
            bb.putInt(0xFFFFFFFF);
        }
        return bb.array();
    }

    /**
     * 设置OTA文件数据
     * @param otaData OTA数据
     */
    public void setOtaData(@NonNull final byte[] otaData) {
        dataProvider = new OtaDataProvider(otaData);
        requestGenerator.setDataProvider(dataProvider);
        // 计算Hash
        hashData = getHash(otaData);
        otaDataSize = otaData.length;
    }

    /**
     * 是否正在进行升级
     * @return 是否正在升级
     */
    public boolean isUpdating() {
        return isUpdating;
    }

    /**
     * 获取开始升级的地址
     * @return 升级地址
     */
    public int getStartAddress() {
        return dataProvider.getStartAddress();
    }

    /**
     * 设置开始升级的地址
     * @param address 升级地址
     */
    public void setStartAddress(int address) {
        dataProvider.setStartAddress(address);
    }

    /**
     * 设置块大小，固件接收一个块后会进行校验并做下一步处理，随后会回复状态到App。
     * @param size 块大小
     */
    public void setBlockSize(int size) {
        dataProvider.setBlockSize(size);
    }

    /**
     * 设置发送一包的最大大小（被MTU大小限制）。
     * @param size 最大包大小
     */
    public void setPacketSize(int size) {
        dataProvider.setPacketSize(size);
    }

    /**
     * 传入OTA固件版本
     * @param firmwareVersion 固件版本
     */
    public void setFirmwareVersion(int firmwareVersion) {
        this.firmwareVersion = firmwareVersion;
    }

    /**
     * 传入固件HASH
     * @param hashData HASH, 4bytes
     */
    public void setHashData(byte[] hashData) {
        this.hashData = hashData;
    }

    /* OTA State */

    @Override
    public void onReceiveNotification(Byte state) {
        Log.v(TAG, "OTA state = " + state);
        switch (state) {
            case OtaState.OTA_STATE_OK:
                cancelTimeout();
                if (!isUpdatePause && !dataProvider.isAllDataSent()) {
                    sendBlock();
                }
                break;
            case OtaState.OTA_STATE_FINISH:
                cancelTimeout();
                // DONE
                notifyOnFinish();
                break;
            case OtaState.OTA_STATE_PAUSE:
                cancelTimeout();
                // Pause
                allowedUpdate = false; // 清掉，不然重新开始的时候会接着之前状态发送数据
                isUpdating = false;
                isUpdatePause = true;
                notifyOnPause();
                break;
            case OtaState.OTA_STATE_CONTINUE:
                // Resume
                notifyOnContinue();
                // Continue to update
                prepareToUpdate();
                break;
            default:
                isUpdating = false;
                reset();
                // Check if retry is needed
                if (needRetryOnState(state)) {
                    retryUpdate();
                } else {
                    notifyOnError(state);
                }
        }
    }

    /* Request Delegate */

    private void sendRequest(@NonNull Request request) {
        if (requestDelegate != null) {
            requestDelegate.sendRequest(request);
        }
    }

    private void sendRequest(@NonNull Request request,
                             @NonNull DeviceCommManager.RequestCallback requestCallback) {
        if (requestDelegate != null) {
            requestDelegate.sendRequest(request, requestCallback);
        }
    }

    public interface RequestDelegate {
        void sendRequest(@NonNull Request request);
        void sendRequest(@NonNull Request request,
                         @NonNull DeviceCommManager.RequestCallback requestCallback);
    }

    /* Event Listener */

    /**
     * OTA事件监听器
     */
    public interface EventListener {

        /* OTA状态 */

        /**
         * 通知设备是否允许升级
         * @param allowed 是否已允许
         */
        void onOtaAllowUpdate(boolean allowed);

        /**
         * OTA已开始
         */
        void onOtaStart();

        /**
         * OTA进度
         * @param progress 进度
         */
        void onOtaProgress(int progress);

        /**
         * OTA已完成
         */
        void onOtaFinish();

        /**
         * OTA已暂停
         */
        void onOtaPause();

        /**
         * OTA已继续
         */
        void onOtaContinue();

        /**
         * OTA遇到错误
         * @param errorCode 错误代码，参考OTA文档
         */
        void onOtaError(int errorCode);

    }

    public interface RetryCallback {
        /**
         * Before retrying next update
         */
        default void beforeRetryCallback() {}

        /**
         * After started retry next update
         */
        default void afterRetryCallback() {}
    }
}

File: ios/Classes/FOTA/Bluetooth/OtaManager.swift
//
//  OTAManager.swift
//  AB OTA Demo
//
//  Created by Bluetrum on 2020/12/25.
//

import Foundation
import CoreBluetooth


public protocol OtaRequestDelegate: AnyObject {
    func sendRequest(_ request: Request, requestCompletion: RequestCompletion?)
}

/// `OtaManager`代理，用于监听OTA事件。
public protocol OtaManagerDelegate: AnyObject {
    /// 通知设备是否允许升级
    func onOtaAllowUpdate(_ allowed: Bool)
    /// OTA已开始
    func onOtaStart()
    /**
     OTA进度
     - Parameters:
        - progress: 进度，[0, 1.0]
     */
    func onOtaProgress(_ progress: Float)
    /// OTA已停止
    func onOtaStop()
    /// OTA已全部完成
    func onOtaFinish()
    /// OTA已暂停
    func onOtaPause()
    /// OTA已继续
    func onOtaContinue()
    /**
     OTA遇到错误
     - Parameters:
        - error: 错误，`OTAError`
     */
    func onOtaError(_ error: OTAError)
}

public class OtaManager: NSObject {
    
    /// 发送每一个数据块之后，等待固件回复的默认超时间隔
    public static let DEFAULT_TIMEOUT: Int = 10 // In second, 10s
    /// 固件默认版本
    public static let UNDEFINED_FIRMWARE_VERSION: UInt32 = 0xFFFFFFFF
    
    public weak var requestDelegate: OtaRequestDelegate?
    
    // MARK: - Properties
    
    /// 代理`OtaManagerDelegate`用以接收OTA状态
    public weak var delegate: OtaManagerDelegate?
    /// OTA文件的版本，默认是0xFFFF，即接受全部版本
    public var firmwareVersion: UInt32 = UNDEFINED_FIRMWARE_VERSION
    public var hashData: Data = Data([0xFF, 0xFF, 0xFF, 0xFF])
    
    weak var logger: LoggerDelegate? = DefaultLogger.shared
    
    var dataProvider: OtaDataProvider?
    var requestGenerator: OtaRequestGenerator
    var allowedUpdate: Bool = false
    
    private var blockSize = OtaConstants.DEFAULT_BLOCK_SIZE {
        didSet {
            dataProvider?.blockSize = blockSize
        }
    }
    public var packetSize = OtaConstants.DEFAULT_PACKET_SIZE {
        didSet {
            dataProvider?.packetSize = packetSize
        }
    }
    
    private var isUpdatePause: Bool = false

    public var isUpdating: Bool = false
    
    var disconnectedDueToDeviceError: Bool = false // 报错后断开不再调用onStop, FIXME: Keep it?
    
    private var timeoutCallback: DispatchWorkItem?

    // MARK: - Public API
    
    /// 构造器
    public override init() {
        requestGenerator = OtaRequestGenerator(dataProvider: nil)
        super.init()
    }
    
    /// OTA Manager初始化，在进行OTA升级之前必须进行初始化。
    public func initialize() {
    }
    
    /**
     释放资源，包括断开蓝牙连接。
     通常在升级完成、遇到错误，或者超时的时候，会自动释放资源，但是如果需要中断升级，亦可手动调用。
     */
    public func deinitialize() {
        firmwareVersion = OtaManager.UNDEFINED_FIRMWARE_VERSION
        allowedUpdate = false
        isUpdating = false
        isUpdatePause = false
        cancelTimeout()
        dataProvider?.reset() // 在蓝牙断开的地方调用
    }
    
    public func prepareToUpdate() {
        if dataProvider != nil {
            requireOtaInfo()
        }
    }
    
    /// 开始进行OTA升级
    public func startOTA() {
        if isReadyToUpdate() {
            isUpdating = true
            notifyOnStart()
            sendBlock()
        }
    }
    
    /**
     判断是否已经就绪。
     
     - Returns: 是否已就绪
     */
    public func isReadyToUpdate() -> Bool {
        return dataProvider != nil && allowedUpdate
    }
    
    /**
     设置OTA数据。
     
     - Parameters:
        - otaData: OTA数据
     */
    public func setOtaData(_ otaData: Data) {
        dataProvider = OtaDataProvider(otaData: otaData)
        requestGenerator.dataProvider = dataProvider
        
        hashData = getDataHash(data: otaData)
    }
    
    // MARK: - Notify
    
    private func notifyAllowUpdate(_ allowed: Bool) {
        DispatchQueue.main.async {
            self.delegate?.onOtaAllowUpdate(allowed)
        }
    }
    
    private func notifyOnStart() {
        DispatchQueue.main.async {
            self.delegate?.onOtaStart()
        }
    }
    
    private func notifyOnProgress(_ progress: UInt32) {
        DispatchQueue.main.async {
            self.delegate?.onOtaProgress(Float(progress))
        }
    }
    
    private func notifyOnFinish() {
        DispatchQueue.main.async {
            self.delegate?.onOtaFinish()
        }
    }
    
    private func notifyOnStop() {
        if !disconnectedDueToDeviceError {
            DispatchQueue.main.async {
                self.delegate?.onOtaStop()
            }
        }
        disconnectedDueToDeviceError = false
    }
    
    private func notifyOnPause() {
        DispatchQueue.main.async {
            self.delegate?.onOtaPause()
        }
    }
    
    private func notifyOnContinue() {
        DispatchQueue.main.async {
            self.delegate?.onOtaContinue()
        }
    }
    
    private func notifyOnError(_ error: OTAError) {
        disconnectedDueToDeviceError = true
        DispatchQueue.main.async {
            self.delegate?.onOtaError(error)
        }
        deinitialize()
    }
    
    private func requireOtaInfo() {
        guard let dataProvider = dataProvider else {
            return
        }
        
        let infoRequest = OtaInfoRequest(version: firmwareVersion, hashData: hashData, dataSize: dataProvider.dataSize)
        sendRequest(infoRequest) { [weak self] request, result, timeout in
            
            if timeout {
                self?.notifyOnError(.timeoutWaitingResponse)
            } else {
                if case let otaInfo? = result, let info = otaInfo as? OtaInfo {
                    print("otaInfo: \(String(describing: info))")
                    
                    self?.dataProvider?.startAddress = info.startAddress
                    self?.dataProvider?.blockSize = info.blockSize
                    self?.allowedUpdate = info.allowUpdate
                    
                    // 如果是从暂停中恢复，自动开始升级流程
                    if self != nil, self!.isUpdatePause {
                        self?.isUpdatePause = false
                        self?.startOTA()
                    } else {
                        self?.notifyAllowUpdate(info.allowUpdate)
                    }
                }
            }
        }
    }
    
    // MARK: -
    
    private func canSendNow() -> Bool {
        if let dataProvider = dataProvider {
            return !dataProvider.isBlockSentFinish()
        }
        return false
    }
    
    // MARK: - Request
    
    private func sendOtaStart() {
        if let dataProvider = dataProvider {
            let request = requestGenerator.getOtaStartRequest()
            sendRequest(request)
            
            // Report progress
            let progress = dataProvider.progress
            notifyOnProgress(progress)
        }
    }
    
    private func sendOtaData() {
        if let dataProvider = dataProvider {
            let request = requestGenerator.getOtaSendDataRequest()
            sendRequest(request)
            
            // Report progress
            let progress = dataProvider.progress
            notifyOnProgress(progress)
        }
    }
    
    private func doSendData() {
        // Continue sending data
        while canSendNow() {
            sendOtaData()
        }
        // 发完块的最后一包后，就等待固件回复
        if let isBlockSentFinish = dataProvider?.isBlockSentFinish(), isBlockSentFinish {
            timeoutCallback?.cancel()
            timeoutCallback = DispatchWorkItem { [weak self] in
                self?.handleTimeout()
            }
            DispatchQueue.global().asyncAfter(deadline: .now() + .seconds(OtaManager.DEFAULT_TIMEOUT), execute: timeoutCallback!)
        }
    }
    
    private func sendBlock() {
        sendOtaStart()
        doSendData()
    }
    
    private func handleTimeout() {
       // logger?.e(.otaManager, "Waiting for response timeout")
        notifyOnError(OTAError.timeoutWaitingResponse)
    }
    
    private func cancelTimeout() {
        timeoutCallback?.cancel()
        timeoutCallback = nil
    }
    
    private func getDataHash(data: Data) -> Data {
        return data.md5().subdata(in: 0..<4)
    }
    
}

extension OtaManager {
    
    private func sendRequest(_ request: Request) {
        sendRequest(request, requestCompletion: nil)
    }
    
    private func sendRequest(_ request: Request, requestCompletion: RequestCompletion?) {
        requestDelegate?.sendRequest(request, requestCompletion: requestCompletion)
    }
    
}

extension OtaManager {
    
    public func updateOTAState(state: UInt8) {
        
        if let otaState = OtaState(rawValue: state) {
            switch otaState {
            case .STATE_OK:
                cancelTimeout()
                // If not in pause state, and not done yet, go on sending, until done
                if !isUpdatePause, let dataProvider = dataProvider, !dataProvider.isAllDataSent() {
                    sendBlock()
                }
            case .STATE_DONE:
                cancelTimeout()
                isUpdating = false
                notifyOnFinish()
            case .STATE_PAUSE:
                cancelTimeout()
                // Pause
                allowedUpdate = false
                isUpdating = false
                isUpdatePause = true
                notifyOnPause()
            case .STATE_CONTINUE:
                // Resume
                notifyOnContinue()
                // Continue to update
                prepareToUpdate()
            }
        } else {
            isUpdating = false
            cancelTimeout()
            notifyOnError(.deviceReport(code: state))
        }
    }
    
}
